File: electron/editPlan.js
/**
 * Parse project JSON into EditPlan structure
 */
function buildPlan(projectJsonString) {
  let parsed;
  try {
    parsed = JSON.parse(projectJsonString);
  } catch (e) {
    throw new Error(`Invalid project JSON: ${e.message}`);
  }

  const { id, assets = {}, clips = {}, tracks = {}, canvasNodes = {} } = parsed;

  if (!id) {
    throw new Error('Project JSON missing id field');
  }

  const mainTrack = [];
  const overlayTrack = [];

  // Create a map of clipId -> canvasNode for quick lookup
  const canvasNodeMap = {};
  Object.values(canvasNodes).forEach(node => {
    canvasNodeMap[node.clipId] = node;
  });

  // Process clips by track role
  for (const [trackId, track] of Object.entries(tracks)) {
    if (!track.clipOrder || !Array.isArray(track.clipOrder)) {
      continue;
    }

    for (const clipId of track.clipOrder) {
      const clip = clips[clipId];
      if (!clip) continue;

      const asset = assets[clip.assetId];
      if (!asset) continue;

      // Convert file:// URLs to local paths
      let srcPath = asset.src;
      if (srcPath.startsWith('file://')) {
        srcPath = srcPath.substring(7); // Remove 'file://' prefix
      }

      // Validate clip timing
      if (clip.outMs <= clip.inMs) {
        throw new Error(`Clip ${clipId} has invalid timing: out <= in`);
      }

      const seqClip = {
        srcPath,
        inMs: clip.inMs,
        outMs: clip.outMs,
        startMs: clip.startMs,
        endMs: clip.endMs,
      };

      // Attach asset metadata for aspect ratio preservation (especially for images)
      if (asset.width && asset.height) {
        seqClip.assetWidth = asset.width;
        seqClip.assetHeight = asset.height;
      }

      // Attach canvasNode for overlay tracks (PiP transforms)
      if (track.role === 'overlay' && canvasNodeMap[clipId]) {
        const canvasNode = canvasNodeMap[clipId];
        seqClip.canvasNode = {
          x: canvasNode.x,
          y: canvasNode.y,
          width: canvasNode.width,
          height: canvasNode.height,
          rotation: canvasNode.rotation,
          opacity: canvasNode.opacity,
        };
      }

      // All clips go to mainTrack for now (we'll handle overlays later)
      if (track.role === 'main') {
        mainTrack.push(seqClip);
      } else {
        overlayTrack.push(seqClip);
      }
    }
  }

  // Sort main track by start time
  mainTrack.sort((a, b) => a.startMs - b.startMs);

  // TODO: Implement proper FFmpeg overlay compositing for overlayTrack clips
  // For now, temporarily add overlay tracks to mainTrack to prevent them from disappearing
  // This means they'll be concatenated sequentially rather than composited, but at least they're exported
  // Future: Use FFmpeg overlay filter to composite PiP clips on top of main track
  overlayTrack.sort((a, b) => a.startMs - b.startMs);
  
  // Merge overlay tracks into main track for now (they'll be processed sequentially)
  // In a proper implementation, we'd use FFmpeg overlay filters to composite them
  mainTrack.push(...overlayTrack);
  mainTrack.sort((a, b) => a.startMs - b.startMs);

  // For overlapping clips (clips on different tracks at same time),
  // we'll just include all of them for now and let the export handle compositing
  // In the future, we can add proper multi-track compositing logic here

  return {
    id,
    mainTrack,
    overlayTrack,
  };
}

/**
 * Find the visible clip at a given timestamp
 */
function findVisibleClip(plan, tMs) {
  return plan.mainTrack.find((clip) => clip.startMs <= tMs && tMs < clip.endMs);
}

module.exports = {
  buildPlan,
  findVisibleClip,
};

File: src-tauri/src/edit_plan.rs
use serde::{Deserialize, Serialize};
use std::path::PathBuf;

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct SeqClip {
    pub src_path: PathBuf,
    pub in_ms: u64,
    pub out_ms: u64,
    pub start_ms: u64,
    pub end_ms: u64,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct EditPlan {
    pub id: String,
    pub main_track: Vec<SeqClip>,
    pub overlay_track: Vec<SeqClip>,
}

impl EditPlan {
    pub fn top_visible_clip(&self, t_ms: u64) -> Option<&SeqClip> {
        self.main_track.iter().find(|c| c.start_ms <= t_ms && t_ms < c.end_ms)
    }
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct ProjectJson {
    pub id: String,
    pub assets: std::collections::HashMap<String, ProjectAsset>,
    pub clips: std::collections::HashMap<String, ProjectClip>,
    pub tracks: std::collections::HashMap<String, ProjectTrack>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct ProjectAsset {
    pub id: String,
    pub kind: String,
    pub name: String,
    pub src: String,
    pub duration_ms: Option<f64>, // durationMs in input
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct ProjectTrack {
    pub id: String,
    pub role: String, // "main" | "overlay"
    pub clip_order: Vec<String>, // clipOrder in input
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct ProjectClip {
    pub id: String,
    pub asset_id: String,
    pub track_id: String,
    pub start_ms: u64,
    pub end_ms: u64,
    pub in_ms: u64,
    pub out_ms: u64,
}

pub fn build_plan(project_json: &str) -> Result<EditPlan, String> {
    let parsed: ProjectJson = serde_json::from_str(project_json)
        .map_err(|e| format!("invalid project json: {}", e))?;

    let mut main: Vec<SeqClip> = Vec::new();
    let mut overlay: Vec<SeqClip> = Vec::new();

    // Process clips by track role
    for (track_id, track) in &parsed.tracks {
        for clip_id in &track.clip_order {
            if let Some(clip) = parsed.clips.get(clip_id) {
                if let Some(asset) = parsed.assets.get(&clip.asset_id) {
                    // Convert file:// URLs to local paths
                    let src_path = if asset.src.starts_with("file://") {
                        asset.src.strip_prefix("file://").unwrap().to_string()
                    } else {
                        asset.src.clone()
                    };

                    if clip.out_ms <= clip.in_ms {
                        return Err(format!("clip {} out <= in", clip_id));
                    }

                    let seq = SeqClip {
                        src_path: PathBuf::from(src_path),
                        in_ms: clip.in_ms,
                        out_ms: clip.out_ms,
                        start_ms: clip.start_ms,
                        end_ms: clip.end_ms,
                    };

                    if track.role == "main" {
                        main.push(seq);
                    } else {
                        overlay.push(seq);
                    }
                }
            }
        }
    }

    // Basic validation: no overlaps on main track
    main.sort_by_key(|c| c.start_ms);
    for i in 1..main.len() {
        if main[i - 1].end_ms > main[i].start_ms {
            return Err("overlapping clips on main track (MVP disallows)".to_string());
        }
    }

    Ok(EditPlan {
        id: parsed.id,
        main_track: main,
        overlay_track: overlay,
    })
}

